subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -232,5024 +169,464 @@ static char *sprint_token(const uint8_t *token, uint8_t tkl)
-			"retransmissions.");
-		return -ENOMEM;
+	*obj_field = o_f;
+
+	if (!LWM2M_HAS_PERM(o_f, LWM2M_PERM_W) &&
+	    !lwm2m_engine_bootstrap_override(msg->ctx, &msg->path)) {
+		return -EPERM;
 	}
 
-	ret = coap_pending_init(msg->pending, &msg->cpkt,
-				&msg->ctx->remote_addr,
-				COAP_DEFAULT_MAX_RETRANSMIT);
-	if (ret < 0) {
-		LOG_ERR("Unable to initialize a pending "
-			"retransmission (err:%d).", ret);
+	if (!obj_inst->resources || obj_inst->resource_count == 0U) {
+		return -EINVAL;
 	}
 
-	return ret;
+	return 0;
 }
 
-static void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx,
-			      uint8_t *buf, uint16_t buf_len,
-			      struct sockaddr *from_addr,
-			      udp_request_handler_cb_t udp_request_handler)
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+static bool bootstrap_delete_allowed(int obj_id, int obj_inst_id)
 {
-	struct lwm2m_message *msg = NULL;
-	struct coap_pending *pending;
-	struct coap_reply *reply;
-	struct coap_packet response;
-	int r;
-	uint8_t token[8];
-
-	r = coap_packet_parse(&response, buf, buf_len, NULL, 0);
-	if (r < 0) {
-		LOG_ERR("Invalid data received (err:%d)", r);
-		return;
-	}
+	char pathstr[MAX_RESOURCE_LEN];
+	bool bootstrap_server;
+	int ret;
 
-	(void)coap_header_get_token(&response, token);
-	pending = coap_pending_received(&response, client_ctx->pendings,
-					ARRAY_SIZE(client_ctx->pendings));
-	if (pending && coap_header_get_type(&response) == COAP_TYPE_ACK) {
-		msg = find_msg(pending, NULL);
-		if (msg == NULL) {
-			LOG_DBG("Orphaned pending %p.", pending);
-			coap_pending_clear(pending);
-			return;
+	if (obj_id == LWM2M_OBJECT_SECURITY_ID) {
+		snprintk(pathstr, sizeof(pathstr), "%d/%d/1",
+			 LWM2M_OBJECT_SECURITY_ID, obj_inst_id);
+		ret = lwm2m_engine_get_bool(pathstr, &bootstrap_server);
+		if (ret < 0) {
+			return false;
 		}
 
-		msg->acknowledged = true;
-
-		if (msg->reply == NULL) {
-			/* No response expected, release the message. */
-			lwm2m_reset_message(msg, true);
-			return;
+		if (bootstrap_server) {
+			return false;
 		}
+	}
 
-		/* If the original message was a request and an empty
-		 * ACK was received, expect separate response later.
-		 */
-		if ((msg->code >= COAP_METHOD_GET) &&
-			(msg->code <= COAP_METHOD_DELETE) &&
-			(coap_header_get_code(&response) == COAP_CODE_EMPTY)) {
-			LOG_DBG("Empty ACK, expect separate response.");
-			return;
-		}
+	if (obj_id == LWM2M_OBJECT_DEVICE_ID) {
+		return false;
 	}
 
-	LOG_DBG("checking for reply from [%s]",
-		lwm2m_sprint_ip_addr(from_addr));
-	reply = coap_response_received(&response, from_addr,
-				       client_ctx->replies,
-				       ARRAY_SIZE(client_ctx->replies));
-	if (reply) {
-		msg = find_msg(NULL, reply);
-
-		if (coap_header_get_type(&response) == COAP_TYPE_CON) {
-			r = lwm2m_send_empty_ack(client_ctx,
-						 coap_header_get_id(&response));
-			if (r < 0) {
-				LOG_ERR("Error transmitting ACK");
-			}
-		}
+	return true;
+}
+static int bootstrap_delete(struct lwm2m_message *msg)
```

Comment:
It looks it won't build with bootstrap, this is static, yet called from within `lwm2m_message_handling.c`